### PR TITLE
[SPARK-32730][SQL] Improve LeftSemi and Existence SortMergeJoin right side buffering

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?

LeftSemi and Existence SortMergeJoin should not buffer all matching right side rows when bound condition is empty, this is unnecessary and can lead to performance degradation especially when spilling happens.

### Why are the changes needed?

Performance improvement.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT and TPCDS benchmarks.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -749,6 +749,15 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
         )
       }
 
+      // LEFT SEMI JOIN without bound condition does not use [[ExternalAppendOnlyUnsafeRowArray]]
+      // so should not cause any spill
+      assertNotSpilled(sparkContext, "left semi join") {
```
Comment (peter-toth) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:754]: Without this fix this UT fails.

**Comment by peter-toth** [comment]
cc @cloud-fan, @maropu, @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 2]
```diff
@@ -673,8 +680,29 @@ private[joins] class SortMergeJoinScanner(
...
+  private[this] val bufferedMatches: AppendOnlyUnsafeRowArray = if (bufferFirstOnly) {
+    new AppendOnlyUnsafeRowArray {
+      var buffer: UnsafeRow = null
+
+      override def clear(): Unit = {
+        buffer = null
+      }
+
+      override def add(row: UnsafeRow): Unit = {
+        assert(buffer == null)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:692]: Are you saying that `ExternalAppendOnlyUnsafeRowArray` will do spill even if we only add one row?

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:692]: No, its threshold parameters do work as expected. Just `ExternalAppendOnlyUnsafeRowArray` looked a bit heavy weight for this case when we want to store only one row. But we can also use `new ExternalAppendOnlyUnsafeRowArray(1, 1)` for this case.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:692]: How did you test it if `ExternalAppendOnlyUnsafeRowArray` doesn't spill either?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala [hunk 3]
```diff
@@ -50,7 +58,7 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
     initialSize: Int,
     pageSizeBytes: Long,
     numRowsInMemoryBufferThreshold: Int,
-    numRowsSpillThreshold: Int) extends Logging {
+    numRowsSpillThreshold: Int) extends AppendOnlyUnsafeRowArray with Logging {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:61]: For core component like this, I remember we rarely change its inheritance. It is easily to have performance regression.

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala]: All right, in that case let's drop that new trait and stick to the important part.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 4]
```diff
@@ -673,8 +680,9 @@ private[joins] class SortMergeJoinScanner(
    */
   private[this] var matchJoinKey: InternalRow = _
   /** Buffered rows from the buffered side of the join. This is empty if there are no matches. */
-  private[this] val bufferedMatches =
-    new ExternalAppendOnlyUnsafeRowArray(inMemoryThreshold, spillThreshold)
+  private[this] val bufferedMatches: ExternalAppendOnlyUnsafeRowArray =
+    new ExternalAppendOnlyUnsafeRowArray(if (bufferFirstOnly) 1 else inMemoryThreshold,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:684]: how does this change avoid spilling?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 5]
```diff
@@ -662,7 +668,8 @@ private[joins] class SortMergeJoinScanner(
     bufferedIter: RowIterator,
     inMemoryThreshold: Int,
     spillThreshold: Int,
-    eagerCleanupResources: () => Unit) {
+    eagerCleanupResources: () => Unit,
+    bufferFirstOnly: Boolean) {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:672]: nit: `bufferFirstOnly` -> `matchedBufferFirstOnly`? And, please add `@param`, too.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:672]: `bufferFirstOnly: Boolean = false` to avoid the unnecessary changes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 4 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:684]: We need this change? How about just setting 1 in the caller side? https://github.com/apache/spark/pull/29572/files#diff-2649ee0724bbc5ee6a64d7de308c527eR255

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:684]: There is another call site in existence join: https://github.com/apache/spark/pull/29572/files#diff-2649ee0724bbc5ee6a64d7de308c527eR336

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 5 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:672]: Thanks, fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 4 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Not this change does it, but the one in `bufferMatchingRows()`. This change just avoids creating a buffer larger than 1.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 6]
```diff
@@ -662,7 +661,8 @@ private[joins] class SortMergeJoinScanner(
     bufferedIter: RowIterator,
     inMemoryThreshold: Int,
     spillThreshold: Int,
-    eagerCleanupResources: () => Unit) {
+    eagerCleanupResources: () => Unit,
+    matchedBufferFirstOnly: Boolean = false) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:665]: how about `onlyBufferFirstMatch`?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 6 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Ok, renamed.

**Comment by maropu** [comment]
Ur one more; please update the title/description (this PR is not only for leftsemi, right?), too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 7]
```diff
@@ -156,8 +156,7 @@ case class SortMergeJoinExec(
               RowIterator.fromScala(rightIter),
               inMemoryThreshold,
               spillThreshold,
-              cleanupResources
-            )
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:160]: nit: please avoid the unnecessary changes.

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Sorry, reverted them.

**Comment by peter-toth** [comment]
> Ur one more; please update the title/description (this PR is not only for leftsemi, right?), too.

Updated, thanks.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by peter-toth** [comment]
Thanks for the review @cloud-fan, @maropu, @viirya.

**Comment by gatorsmile** [comment]
@peter-toth Nice fix! Could you share the perf difference when you run the TPC-DS ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 8]
```diff
@@ -251,7 +251,8 @@ case class SortMergeJoinExec(
               RowIterator.fromScala(rightIter),
               inMemoryThreshold,
               spillThreshold,
-              cleanupResources
+              cleanupResources,
+              condition.isEmpty
```
Comment (juliuszsompolski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:255]: Thanks @peter-toth !
I think this could be also added to LeftAnti join, which is also only interested in the existence of a match and doesn't need to buffer them.

**Comment by peter-toth** [comment]
> @peter-toth Nice fix! Could you share the perf difference when you run the TPC-DS ?

Thanks @gatorsmile. Yes, I will try to run some benchmarks for this particular change and share the results.

BTW, I have another PR open that brings ~30% improvement to some of the TPCDS queries: https://github.com/apache/spark/pull/28885

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 8 above)
Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:255]: Thanks @juliuszsompolski, I think you are right. Shall I open a follow-up PR or a different ticket?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:255]: I think a followup PR is fine.

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:255]: I've opened https://github.com/apache/spark/pull/29727
